design/acme-orders-challenges-crd.md

```diff
@@ -0,0 +1,388 @@
+
+The scope of changes to existing types as been purposely kept as small as possible.
+
+We do however require a change to the `status` field of the Certificate resource
+type. This should not be too problematic for users, as the cert-manager controller
+should be able to tolerate this change seamlessly.
+
+```go
+// CertificateStatus defines the observed state of Certificate
+type CertificateStatus struct {
+	Conditions      []CertificateCondition `json:"conditions,omitempty"`
+    ACME            *CertificateACMEStatus `json:"acme,omitempty"`
+    /////// NEW FIELD ///////
+	LastFailureTime *metav1.Time           `json:"lastFailureTime,omitempty"`
+}
+
+// CertificateACMEStatus holds the status for an ACME issuer
+type CertificateACMEStatus struct {
+    /////// REMOVED ALL OLD FIELDS ///////
+
+    /////// NEW FIELD ///////
+	// Order contains details about the current in-progress ACME Order.
+	// If this field is not set, an Order is not in progress.
+	// This field may point to a failed or inactive Order.
+	// It is not sufficient to check for the presence of this field in order to
+	// determine whether an order is in progress.
+	// +optional
+	OrderRef *LocalObjectReference `json:"orderRef,omitempty"`
+}
+```
+
+### Controllers
+
+Two new controllers will be built to reconcile the new `Order` and `Challenge`
+types.
+
+This two controllers will take the majority of the logic that used to exist in
+the `Prepare` function of the ACME issuer.
+
+#### Order controller
+
+When an order is created, all of the spec fields must be defined.
+The Order controller will be responsible for handling the entirety of an ACME
+order.
+The 'spec' describes an order to be created. The controller will keep the 'status'
+fields up to date by periodically syncing with the ACME server.
+
+* The Order controller will attempt to create a new order with the ACME server
+if the `status.url` field is not set. It will copy details of the order back
+onto the Order resoure, to save subsequent calls to the ACME server when not required.
+
+* If the orders 'state' is 'final' (i.e. one of valid, invalid or expired) then
+the controller will take no further action and return. (TODO: should we delete challenge
+resources for this order here too?)
+
+* For each authorization on the Order, the order controller will select a challenge
+type to use to solve the authorization (based on the Issuer config and Order's
+SolverConfig) and then ensure a Challenge resource for each authorization exists.
+
+* The order controller will check the status of the Challenge resources that are
+related to it, and if any of them have changed state, it will 'resync' the order
+status with the ACME server.
+
+* If the order is 'ready', the order controller will attempt to 'finalize' the
+order using the CSR as specified on the Order resource (i.e. `spec.csr`).
+After this finalization has completed, the order status will be updated and the
+order's `status.certificateURL` field will be set.
```

Comment:
How are order finalization errors handled? The order will be polled, the status updated to invalid, and the manager will return?

Reply:
Yes - the Order resource will be marked 'invalid' (which we class as a 'final' state).

The Certificate controller will then make a decision whether to retry the Order resource. Currently, this back-off logic is:

* When the certificates controller observes that the Order has become 'invalid', record the 'LastFailureTime'
* Do not retry the Order until 5 minutes *after* the LastFailureTime (TODO: we should make this back-off exponential, but this reflects what we previously used to do as well, although in a much cleaner and easier to improve manner)
* Once 5m has passed, a new Order can be attempted.

I may increase this 5m backoff to 30m before the release of v0.6.0 however, as users now have the ability to retry the order manually by running `kubectl delete order` on the certificate's order (which they previously did not have the ability to do)

Reply:
We've now increased this delay to 1h 😄